File: python_client/torch_imagenet.py
import json
import time
import redisai as rai
import ml2rt
from skimage import io
from cli import arguments

if arguments.gpu:
    device = rai.Device.gpu
else:
    device = rai.Device.cpu

con = rai.Client(host=arguments.host, port=arguments.port)

pt_model_path = '../models/pytorch/imagenet/resnet50.pt'
script_path = '../models/pytorch/imagenet/data_processing_script.txt'
img_path = '../data/cat.jpg'

class_idx = json.load(open("../data/imagenet_classes.json"))

image = io.imread(img_path)

pt_model = ml2rt.load_model(pt_model_path)
script = ml2rt.load_script(script_path)

out1 = con.modelset('imagenet_model', rai.Backend.torch, device, pt_model)
out2 = con.scriptset('imagenet_script', device, script)
a = time.time()
tensor = rai.BlobTensor.from_numpy(image)
out3 = con.tensorset('image', tensor)
out4 = con.scriptrun('imagenet_script', 'pre_process_3ch', 'image', 'temp1')
out5 = con.modelrun('imagenet_model', 'temp1', 'temp2')
out6 = con.scriptrun('imagenet_script', 'post_process', 'temp2', 'out')
final = con.tensorget('out')
ind = final[0]
print(ind, class_idx[str(ind)])
print(time.time() - a)

File: js_client/torch_imagenet.js
var Jimp = require('jimp')
var Redis = require('ioredis')
var fs = require('fs')

const model_path = '../models/pytorch/imagenet/resnet50.pt'
const script_path = '../models/pytorch/imagenet/data_processing_script.txt'
const redis = new Redis({ parser: 'javascript' });
 
const json_labels = fs.readFileSync('../data/imagenet_classes.json')
const labels = JSON.parse(json_labels)

const image_width = 224;
const image_height = 224;

async function load_model() {
 

  const model = fs.readFileSync(model_path, {'flag': 'r'})
  const script = fs.readFileSync(script_path, {'flag': 'r'})
  
  redis.call('AI.MODELSET', 'imagenet_model', 'TORCH', 'CPU', "BLOB", model)
  redis.call('AI.SCRIPTSET', 'imagenet_script', 'CPU', "SOURCE", script)
}

async function run(filename) {
  let image = await Jimp.read(filename);
  let input_image = image.cover(image_width, image_height);
  let buffer = Buffer.from(input_image.bitmap.data);

  redis.call('AI.TENSORSET', 'image1',
             'UINT8', image_height, image_width, 4,
             'BLOB', buffer)
 
  redis.call('AI.SCRIPTRUN', 'imagenet_script', 'pre_process_4ch',
             'INPUTS', 'image1', 'OUTPUTS', 'temp1')
 
  redis.call('AI.MODELRUN', 'imagenet_model',
             'INPUTS', 'temp1', 'OUTPUTS', 'temp2')
 
  redis.call('AI.SCRIPTRUN', 'imagenet_script', 'post_process',
             'INPUTS', 'temp2', 'OUTPUTS', 'out')
 
  let out = await redis.call('AI.TENSORGET', 'out', 'VALUES')
  
  let idx = out[2][0]

  console.log(idx, labels[idx])
}

exports.load_model = load_model
exports.run = run


load_model()
const img_path = '../data/cat.jpg'
run(img_path)
